File: skills/weaver-oa-dev/scripts/api_search.js
#!/usr/bin/env node
/**
 * 泛微OA (E-Cology 9) 538 个后端 REST API 快速检索工具 (JavaScript / Node.js 版)
 * 支持按关键词（中文名、URL路径、参数名、模块名）进行精确或模糊检索
 * 运行方式:
 *   node scripts/api_search.js 待办
 *   node scripts/api_search.js -m 人力资源 -X POST
 */

const fs = require('fs');
const path = require('path');

const RAW_JSON_PATH = path.resolve('C:/Users/Admin/.gemini/antigravity/brain/86c241ed-45ea-453f-a153-6d579096f1df/scratch/all_apis_raw.json');

function loadApis() {
  if (fs.existsSync(RAW_JSON_PATH)) {
    const raw = fs.readFileSync(RAW_JSON_PATH, 'utf8');
    return JSON.parse(raw);
  }
  console.error('错误: 找不到 API 数据源文件 ' + RAW_JSON_PATH);
  return [];
}

function searchApis({ keyword = null, moduleName = null, method = null, limit = 20 }) {
  const docs = loadApis();
  const kw = keyword ? keyword.toLowerCase().trim() : null;
  const mod = moduleName ? moduleName.toLowerCase().trim() : null;
  const meth = method ? method.toUpperCase().trim() : null;

  const results = [];

  for (const item of docs) {
    const data = item.data || {};
    const leaf = item.leaf || {};

    const title = data.title || leaf.name || '';
    const url = data.url || '';
    const category = leaf.category || '';
    const fullPath = leaf.full_path || '';
    const m = (data.method || data.methodSelect || (url.includes('paService') ? 'POST' : 'GET')).toUpperCase();
    const memo = data.memo || '';
    const params = (data.params || []).map((p) => p.name || '');

    if (mod && !category.toLowerCase().includes(mod) && !fullPath.toLowerCase().includes(mod)) {
      continue;
    }
    if (meth && meth !== m) {
      continue;
    }
    if (kw) {
      const match =
        title.toLowerCase().includes(kw) ||
        url.toLowerCase().includes(kw) ||
        memo.toLowerCase().includes(kw) ||
        params.some((p) => p.toLowerCase().includes(kw));
      if (!match) {
        continue;
      }
    }

    results.push({
      title,
      url,
      method: m,
      module: category,
      path: fullPath,
      paramsCount: params.length,
      memo
    });
  }

  console.log(`\n共检索到 ${results.length} 条匹配接口 (展示前 ${Math.min(limit, results.length)} 条):\n`);
  console.log(`${'序号'.padEnd(4)} | ${'模块'.padEnd(8)} | ${'Method'.padEnd(6)} | ${'接口名称'.padEnd(30)} | Endpoint 路径`);
  console.log('-'.repeat(100));

  results.slice(0, limit).forEach((r, idx) => {
    const indexStr = String(idx + 1).padEnd(4);
    const modStr = r.module.padEnd(8);
    const methStr = r.method.padEnd(6);
    const titleStr = (r.title.length > 28 ? r.title.slice(0, 25) + '...' : r.title).padEnd(30);
    console.log(`${indexStr} | ${modStr} | ${methStr} | ${titleStr} | ${r.url}`);
  });

  console.log('-'.repeat(100));
  if (results.length > limit) {
    console.log(`提示: 还有 ${results.length - limit} 条结果未展示，可通过指定更多检索条件缩小范围。\n`);
  }
}

// 参数解析
function main() {
  const args = process.argv.slice(2);
  let keyword = null;
  let moduleName = null;
  let method = null;
  let limit = 20;

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg === '-m' || arg === '--module') {
      moduleName = args[++i];
    } else if (arg === '-X' || arg === '--method') {
      method = args[++i];
    } else if (arg === '-n' || arg === '--limit') {
      limit = parseInt(args[++i], 10) || 20;
    } else if (!arg.startsWith('-')) {
      keyword = arg;
    }
  }

  if (!keyword && !moduleName && !method) {
    console.log('泛微OA E9 API 快速检索工具 (Node.js 版)');
    console.log('用法:');
    console.log('  node scripts/api_search.js <关键词>');
    console.log('  node scripts/api_search.js -m <模块名> [-X GET|POST]');
    console.log('\n示例:');
    console.log('  node scripts/api_search.js 待办');
    console.log('  node scripts/api_search.js -m 知识管理');
    console.log('  node scripts/api_search.js /api/hrm');
    return;
  }

  searchApis({ keyword, moduleName, method, limit });
}

main();

File: skills/weaver-oa-dev/scripts/api_search.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
泛微OA (Ecology 9) 后端 REST API 快速检索工具 (Python 版)
"""

import sys
import os
import json
import argparse
import io

# 强制 UTF-8 输出
if sys.platform == 'win32':
    sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8')

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
API_DICT_PATH = os.path.join(SCRIPT_DIR, "apis_dictionary.json")

def load_apis():
    if os.path.exists(API_DICT_PATH):
        with open(API_DICT_PATH, 'r', encoding='utf-8') as f:
            return json.load(f)
    print(f"错误: 找不到 API 数据文件 {API_DICT_PATH}")
    return []

def search_apis(keyword=None, module=None, method=None, limit=20):
    apis = load_apis()
    results = []
    
    kw = keyword.lower() if keyword else None
    mod = module.lower() if module else None
    meth = method.upper() if method else None

    for api in apis:
        api_mod = api.get('module', '')
        api_name = api.get('apiName', '')
        api_path = api.get('path', '')
        api_meth = api.get('method', '')

        if mod and mod not in api_mod.lower():
            continue
        if meth and api_meth.upper() != meth:
            continue
        if kw:
            if kw not in api_name.lower() and kw not in api_path.lower() and kw not in api_mod.lower():
                continue
        results.append(api)

    print(f"\n共检索到 {len(results)} 个匹配接口 (展示前 {min(limit, len(results))} 条):\n")
    print(f"{'序号':<4} | {'模块':<10} | {'方法':<6} | {'接口路径':<52} | 接口名称")
    print("-" * 105)

    for idx, r in enumerate(results[:limit], 1):
        print(f"{idx:<4} | {r.get('module', ''):<10} | {r.get('method', ''):<6} | {r.get('path', ''):<52} | {r.get('apiName', '')}")

    print("-" * 105)
    if len(results) > limit:
        print(f"提示: 还有 {len(results) - limit} 条结果未展示，可通过缩小检索范围查看。\n")
    else:
        print("")

def main():
    parser = argparse.ArgumentParser(description="泛微OA REST API 快速检索工具")
    parser.add_argument("keyword", nargs="?", default=None, help="搜索关键词 (支持接口名、路径、模块)")
    parser.add_argument("-m", "--module", default=None, help="按模块过滤 (如: 工作流程, 人力资源, 知识管理)")
    parser.add_argument("-X", "--method", default=None, help="按 HTTP 方法过滤 (GET, POST, PUT, DELETE)")
    parser.add_argument("-n", "--limit", type=int, default=20, help="限制显示条数 (默认20)")

    args = parser.parse_args()

    if not args.keyword and not args.module and not args.method:
        parser.print_help()
        return

    search_apis(args.keyword, args.module, args.method, args.limit)

if __name__ == '__main__':
    main()
